Dockerfile

```diff
@@ -11,56 +26,75 @@ RUN ln -snf /bin/bash /bin/sh
 #  - git for downloading RMG respoitories
 #  - wget for downloading conda install script
 #  - libxrender1 required by RDKit
-RUN apt-get update
-RUN apt-get install -y make gcc wget git g++ libxrender1
-
-# Install conda
-RUN wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh 
-RUN bash Miniconda3-latest-Linux-x86_64.sh -b -p /miniconda 
-RUN rm Miniconda3-latest-Linux-x86_64.sh
-ENV PATH="$PATH:/miniconda/bin"
+RUN apt-get update && apt-get install -y \
+    git \
+    gcc \
+    g++ \
+    make \
+    libgomp1\
+    libxrender1 \
+    # Clean up
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
```

Comment:
>**very hacky**

While I appreciate your thoughts and critcisms, let's ensure that our communication in a public forum remains professional. I understand you have concerns and I am more than willing to adjust the changes in this Dockerfile.

> not worth it for ~ 3GB

Correct me if I am wrong but in America, the internet is usually not capped? In other countries, such as Australia and South Africa, we do have a limit on download usage unfortunately. Incredibly frustrating growing up with a 10GB cap on our download usage. Anyway, I figure that if we are to be providing RMG-Py it is better not have a bloated Dockerfile as there are possible users from other countries that do not have the luxury of not having uncapped download usage. Every GB and MB does matter. Addtionally, storage matters. As you rightfully said, a minimal example can generate 10MB+. In my personal opinion, and experience, having that extra bit of space helps.

> delivering users an image that does not follow our own installation instructions

It is simple enough to change the order of some commands in the Dockerfile. I can move the RUNs around if it is important for a user to review the Dockerfile and compare it to the instructions online. However, I imagine they will just be pulling the image.

> wrong package manager

Yes, I am using another package manager because of the speed of solving and less overhead. Since we define the channels in our environment file - I did not see this as a problem. Correct me if I am wrong but on our instructions we inform the user to use `Anaconda3-2023.03-Linux-x86_64.sh`. I am sure one could argue that they're the same in a lot of aspects, but I do think this can be confusing to the user as it does not follow the instructions on our website. Maybe we should change the instructions on the website. Eitherway, I am happy to change the base image to miniconda if that is what you prefer

> not setting environment variables in the same way
> All of this is undocumented, not in the RMG-Py installation instructions, and hacky. Why do we set LD_LIBRARY_PATH, PYTHON, and PYCALL_JL_RUNTIME_PYTHON

It wasn't hacky. I understand you may not be familiar with Julia but it can be somewhat fickle at times about certain environment variables. However, you are right, they did not need to be set in the Dockerfile. I had them left over from when I was trying to get Julia working the first time on T3. I will remove them.

> uses many undocumented environment variables
> Why do also then echo a bunch of variable setting into the activate scripts?

I suppose do not remember but when you first joined the team, you also exprienced issues with installing RMS. Here is the issue where we discussed it - https://github.com/ReactionMechanismGenerator/RMG-Py/issues/2353.  So when that issue was occuring, I had to put those into the activation scripts. It wasn't resolved because we moved to 1.8.5 but rather they were included Julia package we download from the RMG channel. You can see it here - https://github.com/ReactionMechanismGenerator/conda-recipes/blob/rmg-deps/julia/build.sh#LL16C11-L16C11
Ultimately, you are correct, they are not needed anymore ( I was being too cautious by keeping them around). I will remove them

> increases "time to first run" because we no longer build RMS beforehand

That is new information to me. Thanks! I didn't realise that would occur. Usually I save my docker container so any modifications I do to it are not lost. I will change `ENV PYTHONDONTWRITEBYTECODE=true` to `false`. I will have a test to see how it goes.

> is twice as long as the current file, which is a maintainance burdenis twice as long as the current file, which is a maintainance burden

I think this 'twice as long' is not really problematic to maintain.

Reply:
> > **very hacky**
> 
> While I appreciate your thoughts and critcisms, let's ensure that our communication in a public forum remains professional. I understand you have concerns and I am more than willing to adjust the changes in this Dockerfile.

Apologies for the rude wording. In software engineering we use "hacky" to refer to code that just "gets the job done" without being particularly elegant. The lack of inline documentation and carrying-forward of unneesecary code is a hallmark of "hacky" code. Please don't interpret this as a criticism of you, it's just a comment about the code.

> 
> > not worth it for ~ 3GB
> 
> Correct me if I am wrong but in America, the internet is usually not capped? In other countries, such as Australia and South Africa, we do have a limit on download usage unfortunately. Incredibly frustrating growing up with a 10GB cap on our download usage. Anyway, I figure that if we are to be providing RMG-Py it is better not have a bloated Dockerfile as there are possible users from other countries that do not have the luxury of not having uncapped download usage. Every GB and MB does matter. Addtionally, storage matters. As you rightfully said, a minimal example can generate 10MB+. In my personal opinion, and experience, having that extra bit of space helps.
> 

Yes, we have the privilege of uncapped network usage (though I also remember growing up on a limited plan).

If our users are strapped for downloads Docker isn't the best option. Surely a source install would be more efficient.

> > delivering users an image that does not follow our own installation instructions
> 
> It is simple enough to change the order of some commands in the Dockerfile. I can move the RUNs around if it is important for a user to review the Dockerfile and compare it to the instructions online. However, I imagine they will just be pulling the image.

The point is the maintenance burden. If we tell users/developers to install RMG one way, and then the Docker that we push _most_ users towards is done another way, we end up maintaining two sets of installation instructions. By following the same procedure in both places, we rule out the possibility of "maybe micromamba resolved the environment wrong" or "maybe this environment variable was set differently".

> 
> > wrong package manager
> 
> Yes, I am using another package manager because of the speed of solving and less overhead. Since we define the channels in our environment file - I did not see this as a problem. Correct me if I am wrong but on our instructions we inform the user to use `Anaconda3-2023.03-Linux-x86_64.sh`. I am sure one could argue that they're the same in a lot of aspects, but I do think this can be confusing to the user as it does not follow the instructions on our website. Maybe we should change the instructions on the website. Eitherway, I am happy to change the base image to miniconda if that is what you prefer

This is a good point. This should be updated.

The instructions also have the user switch the solver to `libmamba`, which achieves the same result as using `mamba` or `micromamba` in terms of solving the environment and downloading the dependencies. Goal of doing this rather than just moving to mamba is for (1) compatibility with experienced users who already have conda (whereas few have mamba) and (2) to take advantage of the more mature conda software.

> 
> > not setting environment variables in the same way
> > All of this is undocumented, not in the RMG-Py installation instructions, and hacky. Why do we set LD_LIBRARY_PATH, PYTHON, and PYCALL_JL_RUNTIME_PYTHON
> 
> It wasn't hacky. I understand you may not be familiar with Julia but it can be somewhat fickle at times about certain environment variables. However, you are right, they did not need to be set in the Dockerfile. I had them left over from when I was trying to get Julia working the first time on T3. I will remove them.

👍 

> 
> > uses many undocumented environment variables
> > Why do also then echo a bunch of variable setting into the activate scripts?
> 
> I suppose you do not remember but when you first joined the team, you also exprienced issues with installing RMS. Here is the issue where we discussed it - #2353. So when that issue was occuring, I had to put those into the activation scripts. It wasn't resolved because we moved to 1.8.5 but rather they were included Julia package we download from the RMG channel. You can see it here - https://github.com/ReactionMechanismGenerator/conda-recipes/blob/rmg-deps/julia/build.sh#LL16C11-L16C11 Ultimately, you are correct, they are not needed anymore ( I was being too cautious by keeping them around). I will remove them

👍 

> 
> > increases "time to first run" because we no longer build RMS beforehand
> 
> That is new information to me. Thanks! I didn't realise that would occur. Usually I save my docker container so any modifications I do to it are not lost. I will change `ENV PYTHONDONTWRITEBYTECODE=true` to `false`. I will have a test to see how it goes.

👍 

> 
> > is twice as long as the current file, which is a maintainance burdenis twice as long as the current file, which is a maintainance burden
> 
> I think this 'twice as long' is not really problematic to maintain.

My issue is not the length itself, per se, but the content. With the unnecessary code removed this would likely no longer be an issue.